Translated : Calculator Project (#32)
kamp-us/turkce-odin-project: temel_bilgiler/javascript_temelleri/hesap_makinesi_projesi.md

### Giriş

> Başardınız! Bu noktada JavaScript'in temellerini _gerçekten_ sağlam bir şekilde kavramış olmalısınız. Tabii ki öğrenilecek daha çok şey var, ancak şu an oldukça fazla şey oluşturabilmeniz gerekmektedir. Son projemiz, şimdiye kadar öğrendiklerinizi bir araya getirecek: JavaScript, HTML ve CSS kullanarak bir ekran üzerinde hesap makinesi yapacaksınız.

Reviewer: ```suggestion
Başardınız! Bu noktada JavaScript'in temellerini _gerçekten_ sağlam bir şekilde kavramış olmalısınız. Tabii ki öğrenilecek daha çok şey var ancak şu an oldukça fazla şey oluşturabiliyor olmalısınız. Son projemiz, şimdiye kadar öğrendiklerinizi bir araya getirecek: JavaScript, HTML ve CSS kullanarak bir ekran üzerinde hesap makinesi yapacaksınız.
```

> Her zamanki gibi bu proje de sizin için kolay olmayacak unsurlar var, ancak kursu şimdiye kadar takip ettiyseniz kesinlikle bitirmek için ihtiyacınız olan her şeye sahipsiniz. Size çeşitli adımları nasıl uygulayacağınızı göstereceğiz ancak yine de bunu nasıl uygulayacağınız size kalmış!

Reviewer: ```suggestion
Her zamanki gibi bu proje de sizin için kolay olmayacak unsurlar var ancak kursu şimdiye kadar takip ettiyseniz kesinlikle bitirmek için ihtiyacınız olan her şeye sahipsiniz. Size çeşitli adımları nasıl uygulayacağınızı göstereceğiz ancak yine de bunu nasıl uygulayacağınız size kalmış!
```

### Uyarı

<div class="lesson-note" markdown="1">

> Bu hesap makinesi projesine başlamadan önce bir uyarıda bulunmamız gerekiyor. JavaScript'te karmaşık matematiksel ifadeleri nasıl değerlendireceğinizi araştırırken, muhtemelen cezbedici [`eval()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval) fonksiyonuyla karşılaşacaksınız. Ancak, bu fonksiyon çok tehlikeli olabilir ve [asla kullanılmamalıdır](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!)! Bu projenin bir parçası olarak ifadeleri hesaplayacak kendi fonksiyonlarınızı oluşturmanız gerekecektir. Aynı `eval()` ile benzer şekilde bu proje için de ifadelerin nasıl hesaplanacağını araştırırken, bir dizeyi değerlendiren bir `new Function()` döndürmenizi öneren çözümlerle karşılaşabilirsiniz. Aynı `eval()` gibi, bu da [güvenli olmayan verilerin değerlendirilmesindeki potansiyel tehlikeler nedeniyle](https://stackoverflow.com/questions/4599857/are-eval-and-new-function-the-same-thing) kullanılmamalıdır. Ayrıca, tüm işi sizin için yapan çözümlerin nerede eğlencesi var? Haydi başlayalım!

Reviewer: ```suggestion
Bu hesap makinesi projesine başlamadan önce bir uyarıda bulunmamız gerekiyor. JavaScript'te karmaşık matematiksel ifadeleri nasıl değerlendireceğinizi araştırırken, muhtemelen cezbedici [`eval()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval) fonksiyonuyla karşılaşacaksınız. Ancak bu fonksiyon çok tehlikeli olabilir ve [asla kullanılmamalıdır](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!)! Bu projenin bir parçası olarak ifadeleri hesaplayacak kendi fonksiyonlarınızı oluşturmanız gerekecektir. Aynı `eval()` ile benzer şekilde bu proje için de ifadelerin nasıl hesaplanacağını araştırırken, bir dizeyi değerlendiren bir `new Function()` döndürmenizi öneren çözümlerle karşılaşabilirsiniz. Aynı `eval()` gibi, bu da [güvenli olmayan verilerin değerlendirilmesindeki potansiyel tehlikeler nedeniyle](https://stackoverflow.com/questions/4599857/are-eval-and-new-function-the-same-thing) kullanılmamalıdır. Ayrıca, tüm işi sizin için yapan çözümlerin nesi eğlenceli? Haydi başlayalım!
```

    - **Hesap makineniz aynı anda tek bir sayı çiftinden fazlasını değerlendirmemelidir.** Örnek: bir sayı butonuna (`12`) ardından bir operatör butonuna (`+`) ikinci bir sayı butonuna (`7`) ve son olarak ikinci bir operatör butonuna (`-`) basıyorsunuz. Hesap makineniz şunları yapmalıdır: ilk olarak, ilk sayı çiftini (`12 + 7`) değerlendirmeli, ikinci olarak bu hesaplamanın sonucunu (`19`) görüntülemeli ve son olarak bu sonucu(`19`) bir sonraki operatörle (`-`) birlikte yeni hesaplamanızdaki ilk sayı olarak kullanmalıdır.
    - Ekranı taşırmamaları için uzun ondalıklı cevapları yuvarlamalısınız.
    - Tüm sayıları veya bir operatörü girmeden önce `=` butonuna basmak sorunlara neden olabilir!
    - "sil" butonu mevcut tüm verileri silmelidir... kullanıcının "sil" butonuna bastıktan sonra gerçekten yeni bir başlangıç yaptığından emin olun
    - Kullanıcı bir sayıyı 0'a bölmeye çalışırsa alaycı bir hata mesajı görüntüleyin... ve hesap makinenizi çökertmesine izin vermeyin!

#### Bonus Ödevler

> - Kullanıcılar gerekli matematiği yaparlarsa ondalıklı sayılar alabilirler, ancak henüz ondalıklı sayıları yazamazlar. Bir `.` butonu ekleyin ve kullanıcıların ondalıklı sayı girmesine izin verin! Ancak birden fazla yazmalarına izin vermediğinizden emin olun: `12.3.56.5.` Bu sayılar üzerinde matematik işlemleri yapmak zordur. \(ekranda zaten bir tane varsa ondalık butonunu devre dışı bırakın\)

Reviewer: ```suggestion
- Kullanıcılar gerekli matematiği yaparlarsa ondalıklı sayılar alabilirler ancak henüz ondalıklı sayıları yazamazlar. Bir `.` butonu ekleyin ve kullanıcıların ondalıklı sayı girmesine izin verin! Ancak birden fazla yazmalarına izin vermediğinizden emin olun: `12.3.56.5.` Bu sayılar üzerinde matematik işlemleri yapmak zordur. \(ekranda zaten bir tane varsa ondalık butonunu devre dışı bırakın\)
```
